src/Functions/fromReadableSize.cpp

```diff
@@ -0,0 +1,265 @@
+
+        size_t getNumberOfArguments() const override { return 1; }
+
+        DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+        {
+            if (arguments.empty())
+                throw Exception(
+                    ErrorCodes::TOO_FEW_ARGUMENTS_FOR_FUNCTION,
+                    "Number of arguments for function {} doesn't match: passed {}, should be 1.",
+                    getName(),
+                    arguments.size());
+
+            if (arguments.size() > 1)
+                throw Exception(
+                    ErrorCodes::TOO_MANY_ARGUMENTS_FOR_FUNCTION,
+                    "Number of arguments for function {} doesn't match: passed {}, should be 1.",
+                    getName(),
+                    arguments.size());
+
+            const IDataType & type = *arguments[0];
+
+            if (!isString(type))
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Cannot format {} as time string.", type.getName());
+
+            return std::make_shared<DataTypeUInt64>();
+        }
+
+        bool useDefaultImplementationForConstants() const override { return true; }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+        {
+            auto col_to = ColumnUInt64::create();
+            auto & res_data = col_to->getData();
+
+            for (size_t i = 0; i < input_rows_count; ++i)
+            {
+                std::string_view str{arguments[0].column->getDataAt(i)};
+                Int64 token_tail = 0;
+                Int64 token_front = 0;
+                Int64 last_pos = str.length() - 1;
+                UInt64 result = 0;
+
+                /// ignore '.' and ' ' at the end of string
+                while (last_pos >= 0 && (str[last_pos] == ' ' || str[last_pos] == '.'))
+                    --last_pos;
+
+                /// no valid characters
+                if (last_pos < 0)
+                {
+                    throw Exception(
+                        ErrorCodes::BAD_ARGUMENTS,
+                        "Invalid expression for function {}, don't find valid characters, str: \"{}\".",
+                        getName(),
+                        String(str));
+                }
+
+                /// last pos character must be character and not be separator or number after ignoring '.' and ' '
+                if (!isalpha(str[last_pos]))
+                {
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Invalid expression for function {}, str: \"{}\".", getName(), String(str));
+                }
+
+                /// scan spaces at the beginning
+                scanSpaces(str, token_tail, last_pos);
+                token_front = token_tail;
+                /// scan unsigned integer
+                if (!scanUnsignedInteger(str, token_tail, last_pos))
```

Comment:
Consider calling one of the existing integer / float parsing functions in src/IO/ReadHelpers.h. They are battle-tested and optimized to death.

Reply:
Done, thanks!

I've finally gone with `tryReadFloatTextPrecise` as `tryReadFloatText` was returning `true` on random string inputs.
Something odd I've found is that both functions (and their non-try flavours) don't return any kind of error when trying to scan a string starting with whitespace, they simply return and don't advance the buffer cursor.

Is this expected behaviour?

Reply:
Messing with the low-level float parsing code seems quite error-prone. I can check it separately but chances are that it is so established that changing the semantics will break existing use cases :-(

Reply:
Absolutely, was just curious about the difference. Adding an explicit check for leading whitespace and using `tryReadFloatTextPrecise` seems to provide the behaviour that we want so we should be good. Thanks!